src/logic/safe/transactions/offchainSigner/EIP712Signer.js

```diff
@@ -0,0 +1,92 @@
+// @flow
+import { EMPTY_DATA } from '~/logic/wallets/ethTransactions'
+import { getWeb3 } from '~/logic/wallets/getWeb3'
+
+const EIP712_NOT_SUPPORTED_ERROR_MSG = "EIP712 is not supported by user's wallet"
+
+const generateTypedDataFrom = async ({
+  baseGas,
+  data,
+  gasPrice,
+  gasToken,
+  nonce,
+  operation,
+  refundReceiver,
+  safeAddress,
+  safeTxGas,
+  to,
+  valueInWei,
+}) => {
+  const typedData = {
+    types: {
+      EIP712Domain: [
+        {
+          type: 'address',
+          name: 'verifyingContract',
+        },
+      ],
+      SafeTx: [
+        { type: 'address', name: 'to' },
+        { type: 'uint256', name: 'value' },
+        { type: 'bytes', name: 'data' },
+        { type: 'uint8', name: 'operation' },
+        { type: 'uint256', name: 'safeTxGas' },
+        { type: 'uint256', name: 'baseGas' },
+        { type: 'uint256', name: 'gasPrice' },
+        { type: 'address', name: 'gasToken' },
+        { type: 'address', name: 'refundReceiver' },
+        { type: 'uint256', name: 'nonce' },
+      ],
+    },
+    domain: {
+      verifyingContract: safeAddress,
+    },
+    primaryType: 'SafeTx',
+    message: {
+      to,
+      value: valueInWei,
+      data,
+      operation,
+      safeTxGas,
+      baseGas,
+      gasPrice,
+      gasToken,
+      refundReceiver,
+      nonce: Number(nonce),
+    },
+  }
+
+  return typedData
+}
+
+type EIP712RpcCallVersion = 'v3'
+
+export const getEIP712Signer = (version: ?EIP712RpcCallVersion) => async txArgs => {
+  const web3 = getWeb3()
+  const typedData = await generateTypedDataFrom(txArgs)
+
+  const jsonTypedData = JSON.stringify(typedData)
+  const signedTypedData = {
+    jsonrpc: '2.0',
+    method: version === 'v3' ? 'eth_signTypedData_v3' : 'eth_signTypedData',
```

Comment:
we should probably use v4 (not super important, but to always stick to the latest version metamask has)

Reply:
so should it be `eth_signTypedData_v4` and `eth_signTypedData` or `eth_signTypedData_v3`, `eth_signTypedData_v4` and `eth_signTypedData`?

Reply:
hmmm :D .... good question ... I think I would go for either only `_v3` or `_v4` and `_v3`

maybe you can define a version prefix and just append it to `eth_signTypedData` and create 3 signers (with prefix `_v4`, `_v3` and ``)

But I am also fine with sticking to `_v3` for now.